Hotfix/fix client invalid rssi

morantr opened the pull request:
On commit 873bf9e I've implemented
a mechanism that blocks the start monitoring request for 3 seconds until
the client IP is received.
This solution wan not good enough since on a real-time system.
I encountered situations that the IP has been received after more than 15 seconds,
and in such a case, the monitoring was not started. This has caused
the statistics polling task not to work as well since the monitoring request
is inserting the client to the monitor DB, and without it, the statistics polling task
measurement will not work.

This PR reverts the commit that blocks the monitoring request, and add messages that update 
the IP on the monitor database independently.

Currently, when sending arp with an empty IP address, the client that receives
the request will ignore it since the empty IP string is translated to 0.0.0.0.
Replace the empty IP string to a wildcard IP to make sure that the client will
respond to the arp request. This will solve the ARP measurement issue that occurs before
the client IP has been updated on the monitor database. 


Another issue resolved on this PR is when a client is connected to the AP before prplMesh
has brought up, the controller
receives the auto-generated client connected notification and send start monitoring
request before the monitor is attached. The outcome is that the monitor ignores
the request and the client is not being monitored.
If a client is not monitored it cannot be measured and therefore could not be steered
by optimal path task.

To resolve this issue, add START_MONITORING_RESPONSE message. The association
with success flag so the association handling task could increment its state machine
only after receiving a response with success.

The auto-generated code will be added to the follow-up commit.

vitalybu commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -898,7 +898,16 @@ bool network_utils::arp_send(const std::string &iface, const std::string &dst_ip
+    // being sent to a specific mac address, then only the requested client will answer.
+    uint32_t dst_ip_uint; 
+    if (dst_ip.empty()) {
+        dst_ip_uint = network_utils::uint_ipv4_from_string("255.255.255.255");
```
```suggestion
        dst_ip_uint = 0xFFFFFFFF; // Broadcast IP address
```

vitalybu replied:
We're using a broadcast address (it's not wildcard). Please update the commit message.

vitalybu commented on common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_action.h:
```diff
@@ -261,7 +257,13 @@ enum eActionOp_MONITOR: uint8_t {
     ACTION_MONITOR_CHANNEL_SCAN_ABORT_NOTIFICATION = 0x42,
     ACTION_MONITOR_CHANNEL_SCAN_FINISHED_NOTIFICATION = 0x43,
-    ACTION_MONITOR_ENUM_END = 0x44,
+    ACTION_MONITOR_STEERING_CLIENT_SET_GROUP_REQUEST = 0x50,
+    ACTION_MONITOR_STEERING_CLIENT_SET_GROUP_RESPONSE = 0x51,
+    ACTION_MONITOR_STEERING_CLIENT_SET_REQUEST = 0x52,
+    ACTION_MONITOR_STEERING_CLIENT_SET_RESPONSE = 0x53,
+    ACTION_MONITOR_STEERING_EVENT_CLIENT_ACTIVITY_NOTIFICATION = 0x54,
+    ACTION_MONITOR_STEERING_EVENT_SNR_XING_NOTIFICATION = 0x55,
+    ACTION_MONITOR_ENUM_END = 0x56,
```
Probably should mention in the commit message why you decided to reorder the opcodes...

vitalybu reviewed (changes requested):
Stopped at "controller: Send client new IP notification".
Will continue later.

vitalybu commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -902,11 +902,12 @@ bool network_utils::arp_send(const std::string &iface, const std::string &dst_ip
+    //     dst_ip_uint = network_utils::uint_ipv4_from_string("255.255.255.255");
+    // } else {
+    //    dst_ip_uint = network_utils::uint_ipv4_from_string(dst_ip);
+    // }
```
Don't forget to revert it if all tests will pass

vitalybu commented on controller/src/beerocks/master/tasks/statistics_polling_task.cpp:
```diff
@@ -27,6 +27,7 @@ statistics_polling_task::statistics_polling_task(db &database_, ieee1905_1::Cmdu
 
 void statistics_polling_task::work()
 {
+    return;
```
Don't forget to revert it if all tests will pass

tomereli commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -898,7 +898,16 @@ bool network_utils::arp_send(const std::string &iface, const std::string &dst_ip
     uint8_t packet_buffer[128];
 
-    uint32_t dst_ip_uint = network_utils::uint_ipv4_from_string(dst_ip);
+    // If the destination IP is empty, there is no point sending the arp, therefore replace the
```
I just don't understand the comment "there is no point sending the arp" - you still do send the arp and there is a very good point. I think the comment should be updated to reflect the reason for why the destination IP may be empty.
Also please explain in the comment which clients were tested and answers broadcast ARPs.

morantr replied:
There is no point sending the arp since any client which will receive it will ignore it because the request has ip of 0.0.0.0.
The reason that the IP is empty is irrelevant for this commit. This commit makes the function robust.

arnout replied:
I believe the real issue is that with IP address 0.0.0.0, the client will ignore it - as mentioned in the commit message.

But actually, this makes me think: why do we ever need an IP address? AFAIU the only reason we have these IP addresses is to be able to do this ARP, but the IP addresses are non-standard. So if we can simply use the broadcast address, we really don't ever need an IP, and we can always use the broadcast address.

Probably not for this PR though.

vitalybu replied:
I agree with @tomereli's first point - The comment is indeed not clear enough.
You can write something similar to: "If the STA's IP address is unknown, instead of sending an empty value (0.0.0.0), a broadcast address is used (255.255.255.255)".

As for the second part, I don't see any reason to list any clients at this point. 
The new implementation is much better than the old one, even if not all clients respond to broadcast ARP requests.
It was tested already with several devices, but in the following couple of weeks will be tested with many more.

tomereli commented on common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:
```diff
@@ -1082,8 +1082,7 @@ bool ap_wlan_hal_dwpal::update_vap_credentials(
         std::string entry_mode;
-        if (hostapd_config_get_value(it.second, "mode", entry_mode) &&
-            entry_mode != "ap") {
+        if (hostapd_config_get_value(it.second, "mode", entry_mode) && entry_mode != "ap") {
```
How is this related to the change?

morantr replied:
It doesn't. the Clang-format changed it.

tomereli commented on controller/src/beerocks/master/tasks/association_handling_task.cpp:
```diff
@@ -28,7 +28,6 @@ using namespace son;
 #define BEACON_MEASURE_REQ_TIME_SPAN 3000
 #define BEACON_MEASURE_MAX_ATTEMPTS 3
 #define REQUEST_RSSI_MEASUREMENT_DELAY 5000
-#define GET_CLIENT_IP_TIMEOUT_SEC 3
```
I think it makes more sense to actually do `git revert <commit>` then `git commit --amend` to update with the (great!) explanation you added in the commit message. This way, the commit headline will include the hash of the reverted commit so its easier for searching in the git log.

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:
```diff
@@ -76,6 +76,7 @@ eActionOp_CONTROL:
   ACTION_CONTROL_CLIENT_RX_RSSI_MEASUREMENT_START_NOTIFICATION: 104
   ACTION_CONTROL_CLIENT_RX_RSSI_MEASUREMENT_NOTIFICATION: 105
   ACTION_CONTROL_CLIENT_NO_RESPONSE_NOTIFICATION: 106
+  ACTION_CONTROL_CLIENT_NEW_IP_ADDRESS_NOTIFICATION: 107
```
```suggestion
  ACTION_CONTROL_CLIENT_DHCP_NOTIFICATION: 107
```

vitalybu replied:
I think that "NEW_IP_ADDRESS" is good enough.
Today the only supported IP source is indeed a DHCP server, but maybe in the future we may allow setting the IP address via BML.

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:
```diff
@@ -275,19 +276,14 @@ eActionOp_MONITOR:
-  ACTION_MONITOR_STEERING_EVENT_CLIENT_ACTIVITY_NOTIFICATION: 41
-  ACTION_MONITOR_STEERING_EVENT_SNR_XING_NOTIFICATION: 42
+  ACTION_MONITOR_CLIENT_NO_ACTIVITY_NOTIFICATION: 37
+  ACTION_MONITOR_CLIENT_NEW_IP_ADDRESS_NOTIFICATION: 38
```
```suggestion
  ACTION_MONITOR_CLIENT_DHCP_NOTIFICATION: 38
```

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_control.yaml:
```diff
@@ -252,6 +252,10 @@ cACTION_CONTROL_CLIENT_NO_RESPONSE_NOTIFICATION:
   _type: class
   mac: sMacAddr 
 
+cACTION_CONTROL_CLIENT_NEW_IP_ADDRESS_NOTIFICATION:
```
```suggestion
cACTION_CONTROL_CLIENT_DHCP_NOTIFICATION:
```

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_monitor.yaml:
```diff
@@ -145,6 +145,11 @@ cACTION_MONITOR_CLIENT_LINK_MEASUREMENTS_11K_RESPONSE:
   _type: class
   params: sLinkMeasurementsResponse11k 
 
+cACTION_MONITOR_CLIENT_NEW_IP_ADDRESS_NOTIFICATION:
```
```suggestion
cACTION_MONITOR_CLIENT_DHCP_NOTIFICATION:
```

tomereli commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -2794,16 +2794,16 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
             beerocks_header
                 ->addClass<beerocks_message::cACTION_CONTROL_CLIENT_DHCP_COMPLETE_NOTIFICATION>();
-        if (notification == nullptr) {
+        if (!notification_in) {
```
Should mention these fixes in the commit message (nitpick)

morantr replied:
I changed the variable name from `notification` to 'notification_in` since I had to add new tlv `notification_out`. I seem weird to me to mention it.

tomereli reviewed (approved):
Mainly nitpick comments so approving. Nice one!

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -38,8 +38,8 @@
 // BPL Error Codes
-#include <bpl/bpl_err.h>
 #include <bpl/bpl_cfg.h>
+#include <bpl/bpl_err.h>
```
I guess this is a clang-format effect... weird...

arnout commented on common/beerocks/bcl/source/network/network_utils.cpp:
```diff
@@ -898,7 +898,16 @@ bool network_utils::arp_send(const std::string &iface, const std::string &dst_ip
+    // being sent to a specific mac address, then only the requested client will answer.
+    uint32_t dst_ip_uint;
+    if (dst_ip.empty()) {
+        dst_ip_uint = 0xFFFFFFFF; // "255.255.255.255" equivalent
```
Maybe it's not needed, but I have the habit of putting UL or at least U after the constant. On a 32-bit system this may be undefined behaviour, I think.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -2820,16 +2820,30 @@ bool master_thread::handle_cmdu_control_message(const std::string &src_mac,
+                    LOG(WARNING) << "Client does not have a valid parent hostap on the database";
+                    return true;
+                }
+                son_actions::send_cmdu_to_agent(src_mac, cmdu_tx, database,
```
I am clearly missing something since this is weird... The master receives a DHCP notification from the agent, and it sends back exactly the same info to the agent...

morantr replied:
I agree it seems weird, but we have to do it like this since the DHCP event is getting to the son_slave which controls the backhaul manager which is not necessarily the son_slave the client is connected to. The controller knows to which son_slave the client is connected to so it can start the client monitoring on the right monitor thread.
Besides that, it is more convenient that this functionality is centralized on the controller from DEBUG POV.

arnout reviewed (approved):
Maybe not for this PR, but I'm thinking that with this change, it would be better / more consistent to remove the ipv4 from cACTION_CONTROL_CLIENT_START_MONITORING_REQUEST and cACTION_CONTROL_CLIENT_RX_RSSI_MEASUREMENT_REQUEST. Either the monitor already has received the DHCP update or it hasn't, but that shouldn't be blocking in either case.

morantr committed:
common: Replace empty IP with a wildcard on send_arp

Currently, when sending arp with an empty IP address, the client that receives
the request will ignore it since the empty IP string is translated to 0.0.0.0.
Replace the empty IP string to a wildcard IP to make sure that the client will
response to the arp request.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Send monitoring request without waiting for IP

On commit 873bf9ed6a3fba6322e8eb001071e991bf108bf1 I've implemented
a mechanism that block the start monitoring request for 3 seconds until
the client IP is received.
This solution wan not good enough since on a real-time system.
I encountered situations that the IP has been received after more than 15 seconds,
and in such case, the monitoring was not started. This has caused
the statistics polling task not to work as well since the monitoring request
is inserting the client to the monitor DB, and without it, the statistics polling task
measurement will not work.

Revert the commit that blocks the monitoring request.

In the followup commit, will be a fix that updates the IP on the monitor database
independently.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Add IP update notification messages

Preparative commit.
Add new messages to the tlvf to update a client IP address on the monitor DB.
In the followup commit the controller will send the notification when receiving
DHCP event and the message will be drilled to the relevant monitor.

As part of this commit, I had to change the numbering of the monitor messages to
put the new message with others client messages.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Update auto-generated code

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Send client new IP notification

Send client new IP notification upon receiving the DHCP event on the controller.
Do it instead of initiating the client association_task again by calling to
handle_completed_connection() function which is not necessary anymore,
since the whole purpose of it was to trigger the task again with a valid IP address.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Drill IP notification message to the monitor

Drill the new IP notification message to the monitor, and save it on the
monitor database.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Fix enum typo

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Change disconnect type to enum

The member "type" on CLIENT_DISCONNECT_REQUEST has been changed from uint32_t
to eDisconnectType.
There is no reason to pass it as uint32_t, and therefore changed to enum.
In the followup commits, the auto-generated file and the use of the member will
be updated.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Update auto-generated files and tlvf use

Update auto-generated files and tlvf use on CLIENT_DISCONNECT_REQUEST
from the previous commit.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Change functions argument to a const reference

Change disconnect_client functions argument to a const reference.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Remove client disconnect code duplication

Although a client disconnect function exists on son_action, some places
on the code duplicate its implementation.
Remove code duplication and call the function instead.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Add start monitoring response message

When a client is connected to the AP before prplMesh has brought up, the controller
receives the auto-generated client connected notification and send start monitoring
request before the monitor is attached. The outcome is that the monitor ignores
the request, and the client is not being monitored.
If a client is not monitored it cannot be measured and therefore could not be steered
by optimal path task.

To resolve this issue, add START_MONITORING_RESPONSE message. The association
with success flag so the association handling task could increment its state machine
only after receiving a response with success. If not, print a warning to the log.

The auto-generated code will be added to the follow-up commit.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
tlvf: Add START_MONITORING_RESPONSE auto generated code

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Replace if-else with switch-case

Replace if-else with switch-case on handle_responses_timeout function on
the association handling task.
On one of the followup commits another case is going to be added.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Set task state max attempts before a switch to a state

The "attempts" state counter and max_attempts value of for each state
has not been initialized in the right place which is right before moving to
a task state.
Move the "attempts" and "max_attempts" initialization to right before
moving to a relevant state.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add task id to START_MONITORING_REQUEST

Preparative commit.
Since that on the follow-up commits the monitor will return a response for
START_MONITORING_REQUEST, it is needed to add task id to the request so
the response would get back to the task.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add START_MONITORING_RESPONSE message handling

Add START_MONITORING_RESPONSE message handling to the association handling task.
As part of this commit, response and response timeout handlers to the
START_MONITORING_RESPONSE message were added.
The next state selection has been moved to the response handler, and
some prints that were not in the ideal place has been moved to the correct place.

If a response with success is not being received, the task will print warning
message that leaving connected client unmonitored will cause to invalid
measurement on that client, and lead to a failure steering that client.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Drill START_MONITORING_RESPONSE to the controller

Drill START_MONITORING_RESPONSE from the monitor thread to the agent and
from the agent to the controller.
If the monitor has not been attached, when receiving the request message,
return success=false.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr commented:
Add a few more commits.
The new commits start with the commit:
"tlvf: Fix enum typo" facd076570171e762fb01661189c75faf135aa61

vitalybu commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:
```diff
@@ -70,13 +70,14 @@ eActionOp_CONTROL:
-  ACTION_CONTROL_CLIENT_RX_RSSI_MEASUREMENT_NOTIFICATION: 105
-  ACTION_CONTROL_CLIENT_NO_RESPONSE_NOTIFICATION: 106
-  ACTION_CONTROL_CLIENT_NEW_IP_ADDRESS_NOTIFICATION: 107
+  ACTION_CONTROL_CLIENT_START_MONITORING_RESPONSE: 101
```
> If not, the client will be disconnected.

Wait.. What? We don't want to disconnect already connected clients!
If the monitor is not ready yet, or does not respond within some (pre-defined) timeout, the message should be sent again.

vitalybu replied:
Only by reviewing the following commits, I saw that you actually implemented a retries mechanism.
This should probably be mentioned in the commit message... Still think that disconnecting the station is too harsh.
Left a separate comment there.

tomereli replied:
That's what I thought, planned to leave a comment tomorrow. We can't disconnect a connected client because of our internal logic! Where did the "passive" in passive mode go?

vitalybu replied:
So unless anyone objects, let's just log a warning and keep the client connected.
If we'll get additional reports of un-monitored stations in the future, we'll re-think this solution.

morantr replied:
My thought was that leaving a client unmonitored is bad because it means that the client won't be measured and therefore couldn't be steered. But Ok I'll remove the disconnect and replace it with warning print.

vitalybu commented on controller/src/beerocks/master/tasks/association_handling_task.cpp:
```diff
@@ -29,6 +29,8 @@ using namespace son;
 #define REQUEST_RSSI_MEASUREMENT_MAX_ATTEMPTS 5
 #define REQUEST_RSSI_MEASUREMENT_DELAY 5000
+#define START_MONITORING_RESPONSE_TIMEOUT_MSEC 3000
+#define START_MONITORING_MAX_ATTEMPTS (60 * START_MONITORING_RESPONSE_TIMEOUT_MSEC) // 3 minutes
```
I think that 3 minutes is too long... 30 seconds should be enough.

morantr replied:
The maximum waiting for a single response is 3 seconds. 3 minutes is the maximum timeout for repeated requests which is a reasonable time for the wifi to be up.
I think I'll keep it.

vitalybu commented on controller/src/beerocks/master/tasks/association_handling_task.cpp:
```diff
@@ -256,6 +253,45 @@ void association_handling_task::handle_response(std::string mac,
+            if (attempts >= max_attempts) {
+                TASK_LOG(ERROR) << "state START_RSSI_MONITORING reached maximum attempts="
+                                << attempts << " aborting task, and disconnecting client!";
+                son_actions::disconnect_client(database, cmdu_tx, sta_mac, original_parent_mac,
```
That's quite a drastic move...
The problem is probably not with the station, but with the monitor (not ready yet?).
@tomereli, @arnout, what do you guys think?

tomereli replied:
https://github.com/prplfoundation/prplMesh/pull/789#discussion_r376810230

morantr replied:
I'll remove the disconnect call.
https://github.com/prplfoundation/prplMesh/pull/789#discussion_r377014434

vitalybu commented on controller/src/beerocks/master/tasks/association_handling_task.cpp:
```diff
@@ -256,6 +253,45 @@ void association_handling_task::handle_response(std::string mac,
+
+        if (!response->success()) {
+            ++attempts;
+            if (attempts >= max_attempts) {
```
What's wrong with: `if (++attempts >= max_attempts) {`?

morantr replied:
will change

vitalybu commented on controller/src/beerocks/master/tasks/association_handling_task.cpp:
```diff
@@ -256,6 +253,45 @@ void association_handling_task::handle_response(std::string mac,
+        if (!response->success()) {
+            ++attempts;
+            if (attempts >= max_attempts) {
+                TASK_LOG(ERROR) << "state START_RSSI_MONITORING reached maximum attempts="
```
```suggestion
                TASK_LOG(ERROR) << "state START_RSSI_MONITORING reached maximum retries = "
```

morantr replied:
Not a critic one. I'll prefer staying with "attempts" since it is the variable name

vitalybu commented on controller/src/beerocks/master/tasks/association_handling_task.cpp:
```diff
@@ -364,6 +400,20 @@ void association_handling_task::handle_responses_timeout(
     switch (state) {
+    case START_RSSI_MONITORING: {
+        TASK_LOG(DEBUG) << "response for rx rssi measurement from " << original_parent_mac
+                        << " for sta " << sta_mac << " timed out! attempts=" << attempts;
```
```suggestion
                        << " for sta " << sta_mac << " timed out! retries = " << attempts;
```

morantr replied:
https://github.com/prplfoundation/prplMesh/pull/789#discussion_r377016150

vitalybu commented on controller/src/beerocks/master/tasks/association_handling_task.cpp:
```diff
@@ -364,6 +400,20 @@ void association_handling_task::handle_responses_timeout(
+        TASK_LOG(DEBUG) << "response for rx rssi measurement from " << original_parent_mac
+                        << " for sta " << sta_mac << " timed out! attempts=" << attempts;
+        if (attempts >= max_attempts) {
+            TASK_LOG(ERROR) << "state START_RSSI_MONITORING reached maximum attempts=" << attempts
```
```suggestion
            TASK_LOG(ERROR) << "state START_RSSI_MONITORING reached maximum retries = " << attempts
```

morantr replied:
https://github.com/prplfoundation/prplMesh/pull/789#discussion_r377016150

vitalybu commented on controller/src/beerocks/master/tasks/association_handling_task.cpp:
```diff
@@ -364,6 +400,20 @@ void association_handling_task::handle_responses_timeout(
+    }
     case REQUEST_RSSI_MEASUREMENT: {
         TASK_LOG(DEBUG) << "response for rx rssi measurement from " << original_parent_mac
                         << " for sta " << sta_mac << " timed out! attempts=" << attempts;
```
```suggestion
                        << " for sta " << sta_mac << " timed out! retries = " << attempts;
```

morantr replied:
https://github.com/prplfoundation/prplMesh/pull/789#discussion_r377016150

tomereli reviewed (changes requested):
We just can't disconnect a client which was connected before prplmesh started. Actually, we can't disconnect a client ever unless it's part of active steering, and also in this case it's temporary.
This is the opposite of passive.
